spec/controllers/groceries_controller_spec.rb

```diff
@@ -55,25 +79,82 @@
       }
     }
 
-    it 'should finish the current grocery list' do
-      grocery = user_group.groceries.first
-      expect(grocery.finished?).to eq false
-      subject
-      expect(grocery.reload.finished?).to eq true
+    let(:grocery) { user_group.groceries.first }
+
+    subject {
+      patch :finish,
+      id: user_group.groceries.first.id,
+      finish: finish_params
+    }
+
+    context 'groceries are valid to finish' do
+      it 'should finish the current grocery list' do
+        expect(grocery.finished?).to eq false
+        subject
+        expect(grocery.reload.finished?).to eq true
+      end
+
+      it 'should remove specified items from current list' do
+        current_items = grocery.items.first(1)
+        subject
+        expect(grocery.reload.items.to_a).to eq current_items
+      end
+
+      it 'should add specified items to the next list' do
+        next_items = grocery.items[1..-2]
+        subject
+        new_grocery = user_group.groceries.last
+        expect(new_grocery.reload.items.to_a).to eq next_items
+      end
+
+      it 'should redirect to the newly created grocery' do
+        subject
+        new_grocery = user_group.groceries.last
+        expect(response).to redirect_to new_grocery
+      end
     end
 
-    it 'should remove specified items from current list' do
-      grocery = user_group.groceries.first
-      current_items = grocery.items.first(1)
-      subject
-      expect(grocery.reload.items.to_a).to eq current_items
+    context 'groceries are not both valid' do
+      before :each do
+        finish_params[:name] = ''
+      end
+
+      it 'should not finish the current grocery' do
+        subject
+        expect(grocery.reload.finished_at).to be_nil
+      end
+
+      it 'should not create the new grocery' do
+        expect { subject }.to_not change(Grocery, :count)
+      end
+
+      it 'should notify that there was a problem' do
+        expect(subject).to redirect_to grocery
+      end
     end
+  end
 
-    it 'should add specified items to the next list' do
-      next_items = grocery.items[1..-2]
-      subject
-      new_grocery = user_group.groceries.last
-      expect(new_grocery.reload.items.to_a).to eq next_items
+  describe 'POST email_group' do
+    subject { post :email_group, id: grocery.id }
+
+    it 'should deliver an email to each group member' do
+      users = grocery.user_group.users
+      expect { subject }.to change { ActionMailer::Base.deliveries.count }.by users.count
+    end
+
+    it 'should redirect to the grocery page' do
+      expect(subject).to redirect_to grocery
+    end
+  end
+
+  describe 'GET recipes' do
+    it 'should be successful' do
+      stub_request(:get, /food2fork.com/).
+        with(headers: {'Accept'=>'*/*', 'User-Agent'=>'Ruby'}).
```

Comment:
Surrounding space missing for operator '=>'.<br>Space inside { missing.<br>Space inside } missing.